libparsec/crates/testbed/src/template/events.rs

```diff
@@ -1244,142 +1254,392 @@ impl TestbedEventShareRealm {
+            utils::assert_organization_bootstrapped(&builder.events);
+        }
 
-        let greeter_user_id = utils::assert_organization_bootstrapped(&builder.events)
-            .first_user_device_id
-            .user_id()
+        let author = utils::non_revoked_realm_owners(&builder.events, realm)
+            .next()
+            .expect("At least one owner must be present at anytime")
             .to_owned();
 
         // 2) Actual creation
 
-        let token = builder.counters.next_invitation_token();
+        let participants = utils::non_revoked_realm_members(&builder.events, realm)
+            .map(|(device, _)| {
+                let participant_user_id = device.user_id();
+                let participant_public_key = builder
+                    .events
+                    .iter()
+                    .find_map(|e| match e {
+                        TestbedEvent::BootstrapOrganization(x)
+                            if x.first_user_device_id.user_id() == participant_user_id =>
+                        {
+                            Some(x.first_user_private_key.public_key())
+                        }
+                        TestbedEvent::NewUser(x)
+                            if x.device_id.user_id() == participant_user_id =>
+                        {
+                            Some(x.private_key.public_key())
+                        }
+                        _ => None,
+                    })
+                    .expect("User must exist");
+                (participant_user_id.to_owned(), participant_public_key)
+            })
+            .collect();
+
+        let (key_index, keys) = builder
+            .events
+            .iter()
+            .rev()
+            .find_map(|e| match e {
+                TestbedEvent::RotateKeyRealm(x) if x.realm == realm => {
+                    let mut keys = x.keys.clone();
+                    keys.push(builder.counters.next_secret_key());
+                    Some((x.key_index + 1, keys))
+                }
+                _ => None,
+            })
+            .unwrap_or((1, vec![builder.counters.next_secret_key()]));
+
+        let keys_bundle_access_key = builder.counters.next_secret_key();
+
         Self {
-            created_on: builder.counters.next_timestamp(),
-            greeter_user_id,
-            claimer_email,
-            token,
+            timestamp: builder.counters.next_timestamp(),
+            author,
+            realm,
+            key_index,
+            keys,
+            keys_bundle_access_key,
+            custom_key_canary: None,
+            encryption_algorithm: SecretKeyAlgorithm::Xsalsa20Poly1305,
+            hash_algorithm: HashAlgorithm::Sha256,
+            participants,
+            cache: Arc::default(),
         }
     }
+
+    pub fn certificates<'a: 'c, 'b: 'c, 'c>(
+        &'a self,
+        template: &'b TestbedTemplate,
+    ) -> impl Iterator<Item = TestbedTemplateEventCertificate> + 'c {
```

Comment:
Is it usefull that you have 3 differents lifetime ? Can you provide a comment explaining why ?

Reply:
I've added this comment:
```Rust
    // We need three lifetimes here to describe the fact the output iterator
    // (lifetime 'c) wraps data from both the self object (lifetime 'a)
    // and the template (lifetime 'b). Hence 'c outliving both 'a and 'b.
```